service/src/main/java/bio/terra/workspace/service/workspace/flight/WaitForProjectPermissionsStep.java

```diff
@@ -0,0 +1,89 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.GCP_PROJECT_ID;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.GcpUtils;
+import bio.terra.workspace.common.utils.RetryUtils;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import java.time.Duration;
+import java.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Do not complete the cloud context creation until the project permissions have propagated. */
+public class WaitForProjectPermissionsStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(WaitForProjectPermissionsStep.class);
+  private final AuthenticatedUserRequest userRequest;
+
+  public WaitForProjectPermissionsStep(AuthenticatedUserRequest userRequest) {
+    this.userRequest = userRequest;
+  }
+
+  /**
+   * CRL doesn't support list-bucket. For now we call GCP directly.
+   *
+   * @param userRequest user creating the cloud contxt
+   * @param gcpProjectId project that got created
+   * @return GCS Storage object
+   */
+  private Storage getStorage(AuthenticatedUserRequest userRequest, String gcpProjectId) {
+    StorageOptions.Builder optionsBuilder = StorageOptions.newBuilder();
+    optionsBuilder.setCredentials(GcpUtils.getGoogleCredentialsFromUserRequest(userRequest));
+    optionsBuilder.setProjectId(gcpProjectId);
+    return optionsBuilder.build().getService();
+  }
+
+  /**
+   * Method to be used for testing that project permissions have propagated. We are assuming that if
+   * one permission is working, they all have been sync'd. As far as we know, that is the case.
```

Comment:
From google support on 1/16/23
> Does IAM propagate differently across region?
> - Yes, it's possible different regions can see results at different times. However they should be close enough.
> - Given IAM has been migrating to a new stack from early Nov, it's possible that your workflow may have been>  affected by IAM having to rely on one system for one region and other for another, both systems not fully synced. Given we expect our migration to be completed by next week, this issue should be resolved by then.
> - In case if  you still experiencing cross-region issues, please let me know
> - Aside, if there are any regional outages or issues , such workflows are prone to fail someday and you should be notified of them.

I mention this because the control plane (which is where this check runs) can be in a different region from the resource being tested. I don't know if there is much we can do. This just underscores the complexity.